devito/operator/operator.py

```diff
@@ -458,12 +460,19 @@ def _lower_iet(cls, uiet, profiler=None, **kwargs):
 
         # Lower IET to a target-specific IET
         graph = Graph(iet, **kwargs)
-        graph = cls._specialize_iet(graph, **kwargs)
 
+        exprs = FindNodes(Expression).visit(graph.root)
+        if any(isinstance(func, PETScArray) for expr in exprs for func in expr.functions):
+            lower_petsc(graph, **kwargs)
+
+        graph = cls._specialize_iet(graph, **kwargs)
         # Instrument the IET for C-level profiling
         # Note: this is postponed until after _specialize_iet because during
         # specialization further Sections may be introduced
-        cls._Target.instrument(graph, profiler=profiler, **kwargs)
+        # TODO: Figure out how to include the timers with PETSc
+        if not any(isinstance(func,
+                              PETScArray) for expr in exprs for func in expr.functions):
```

Comment:
Somehow this logic should be moved inside `instrument`, perhaps after being carefully abstracted